demosplan/DemosPlanCoreBundle/DoctrineMigrations/2023/01/Version20230104090531.php

```diff
@@ -0,0 +1,106 @@
+
+use Doctrine\DBAL\Exception;
+use Doctrine\DBAL\Schema\Schema;
+use Doctrine\Migrations\AbstractMigration;
+
+class Version20230104090531 extends AbstractMigration
+{
+    private const NEW_GIS_NAME = 'basemap';
+    private const NEW_GIS_LAYER = 'de_basemapde_web_raster_farbe';
+    private const NEW_GIS_URL = 'https://sgx.geodatenzentrum.de/wms_basemapde';
+    private const NEW_COPYRIGHT_TEXT = '© basemap.de BKG (www.basemap.de) / LVermGeo SH (www.LVermGeoSH.schleswig-holstein.de)';
+
+    private const OLD_GIS_NAME = 'Web-Atlas';
+    private const OLD_GIS_LAYER = 'webatlasde';
+    private const OLD_GIS_URL = 'https://sg.geodatenzentrum.de/wms_webatlasde__ce01ef82-8df3-d28f-edc0-bae62cfa13d6';
+    private const OLD_COPYRIGHT_TEXT = 'Kartengrundlage: © GeoBasis-DE/LVermGeo SH (www.LVermGeoSH.schleswig-holstein.de)';
+
+    public function getDescription(): string
+    {
+        return 'refs T29639: Replace the "Web-Atlas" by basemap as standard and sets a new copyright,
+        only for procedure templates.';
+    }
+
+    /**
+     * @throws Exception
+     */
+    public function up(Schema $schema): void
+    {
+        $this->abortIfNotMysql();
+
+        // Replace all Web-Atlas Entries by basemap (only for procedure templates).
+        $this->addSql(
+            'UPDATE _gis AS g
+            INNER JOIN _procedure AS p ON g._p_id = p._p_id
+            SET
+                g._g_name = :gisName,
+                g._g_layers = :gisLayer,
+                g._g_url = :newGisUrl
+            WHERE g._g_url = :oldGisUrl AND p._p_master = 1 AND p._p_deleted = 0',
+            [
+                'gisName'     => self::NEW_GIS_NAME,
+                'gisLayer'    => self::NEW_GIS_LAYER,
+                'newGisUrl'   => self::NEW_GIS_URL,
+                'oldGisUrl'   => self::OLD_GIS_URL,
+            ]
+        );
+
+        // Update the customers.
+        $this->addSql(
+            'UPDATE customer AS c
+                SET c.base_layer_url = :newGisUrl,
+                    c.base_layer_layers = :newGisLayer',
+            [
+                'newGisUrl'   => self::NEW_GIS_URL,
+                'newGisLayer' => self::NEW_GIS_LAYER,
+            ]
+        );
+
+        // Replace the old copyright entries by the new ones.
+        $this->addSql(
+            'UPDATE _procedure_settings AS ps
+            INNER JOIN _procedure AS p ON ps._p_id = p._p_id
+            SET ps.copyright = :newCopyrightText
+            WHERE ps.copyright = :oldCopyrightText AND p._p_master = 1 AND p._p_deleted = 0',
+            [
+                'newCopyrightText' => self::NEW_COPYRIGHT_TEXT,
+                'oldCopyrightText' => self::OLD_COPYRIGHT_TEXT,
+            ]
+        );
+    }
+
+    /**
+     * @throws Exception
+     */
+    public function down(Schema $schema): void
+    {
+        $this->abortIfNotMysql();
+
+        $this->throwIrreversibleMigrationException('This migration is not revertable because we can\'t
```

Comment:
Please do not throw an exception, just silently do nothing

Reply:
adjusted in [8c4f137](https://github.com/demos-europe/demosplan-core/pull/578/commits/8c4f1371225fbcc1a3f78c282b31755b79778af8)

Reply:
The comment as comment would be helpful :-)

Reply:
of course, you are right :)
done in [952fba1](https://github.com/demos-europe/demosplan-core/pull/578/commits/952fba1e2ec34e5ab178385afab8f2a33eff6c55)

Reply:
Why shouldn't an exception be thrown?

Reply:
There are cases when we need to down migrate several migrations. Consider we are at migration E and want to go back to B for example because an installation failed and you need to fix something in the infrastructure before you could retry the installation again. Of course you could use the backup, but there are cases, where this is complicated or strictly not necessary, as you can quite easily and rapidly fix the issue.
When in this case migration D would throw an exception then the whole process is stopped and there is no way around it.
Down migrations are seldomly used, but when you need them you are really happy if you have them